File: 0114-flatten-binary-tree-to-linked-list/0114-flatten-binary-tree-to-linked-list.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public void flatten(TreeNode root) {
    if(root==null){
       return;
     }
    flatten(root.left);
    flatten(root.right);
    //flatten left and right subtree first
    TreeNode left=root.left;
    TreeNode right=root.right;
    root.right=left;
    root.left=null;
    TreeNode temp=root;
    while(temp.right!=null){
    temp=temp.right;
    }
    temp.right=right;
   }
}

File: 0114-flatten-binary-tree-to-linked-list/0114-flatten-binary-tree-to-linked-list.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void flatten(TreeNode* root) {
    if(root==NULL)return;
    flatten(root->left);
    flatten(root->right);
    TreeNode* leftt=root->left;
    TreeNode* rightt=root->right;
    root->right=leftt;
    root->left=NULL;
    TreeNode* temp=root;
    while(temp->right!=NULL){
        temp=temp->right;
    }
    temp->right=rightt; 
    }
};